tests/integration/integration_suite_test.go

```diff
@@ -1,13 +1,187 @@
+// - Cancel the repair
+// - Terminate the namespace and delete the cluster
+func TestReaperDeploymentScenario(t *testing.T) {
+	t.Run("Initialize cluster", initializeCluster)
+
+	t.Run("Deploy cluster", deployClusterForReaper)
+	t.Run("Check expected resources are present", checkResourcePresenceForReaper)
+	t.Run("Wait for Reaper pod to be ready", waitForReaperPod)
+	t.Run("Check Reaper has registered the required elements", checkReaperRegistered)
+	t.Run("Start repair on the reaper_db keyspace", triggerRepair)
+	t.Run("Wait for at least one segment to be processed and cancel the repair", waitForSegmentDoneAndCancel)
+
+	t.Run("Tear down cluster", cleanupCluster)
+}
+
+func deployClusterForReaper(t *testing.T) {
+	DeployClusterWithValues(t, "default", "three_nodes_cluster_with_reaper.yaml")
+}
+
+func checkResourcePresenceForReaper(t *testing.T) {
+	CheckResourceIsPresent(t, "service", "app.kubernetes.io/managed-by=reaper-operator")
+	CheckClusterExpectedResources(t)
+}
+
+func waitForReaperPod(t *testing.T) {
+	WaitForReaperPod(t)
+}
+
+func checkReaperRegistered(t *testing.T) {
+	CheckKeyspaceExists(t, "reaper_db")
+	CheckClusterIsRegisteredInReaper(t, "k8ssandra")
+}
+
+func triggerRepair(t *testing.T) {
+	TriggerRepair(t, "reaper_db")
+}
+
+func waitForSegmentDoneAndCancel(t *testing.T) {
+	WaitForOneSegmentToBeDone(t)
+	CancelRepair(t)
+}
+
+// Medusa scenario (invoked with a specific backend name):
+// - Register a cluster with 1 node
+// - Potentially install backend specific dependencies (such as Minio)
+// - Create the backend credentials secret
+// - Create a keyspace and a table
+// - Load 10 rows and check that we can read 10 rows
+// - Perform a backup with Medusa
+// - Load 10 rows and check that we can read 20 rows
+// - Restore the backup
+// - Verify that we can read 10 rows
+// - Terminate the namespace and delete the cluster
+func TestMedusaDeploymentScenario(t *testing.T) {
+	backends := []string{"Minio", "S3"}
+	for _, backend := range backends {
+		t.Run(fmt.Sprintf("%s: Initialize cluster", backend), initializeCluster)
+
+		t.Run(fmt.Sprintf("%s: Create medusa secret", backend), func(t *testing.T) {
+			if backend == "Minio" {
+				DeployMinioAndCreateBucket(t, "k8ssandra-medusa")
+				CreateMedusaSecretWithFile(t, "secret/medusa_minio_secret.yaml")
+			} else {
+				// S3
+				CreateMedusaSecretWithFile(t, "~/medusa_secret.yaml")
+			}
+		})
+
+		t.Run(fmt.Sprintf("%s: Deploy cluster", backend), func(t *testing.T) {
+			if backend == "Minio" {
+				// Minio
+				DeployClusterWithValues(t, "minio", "one_node_cluster_with_medusa_minio.yaml")
+			} else {
+				// S3
+				DeployClusterWithValues(t, "default", "one_node_cluster_with_medusa_s3.yaml")
+			}
+		})
+		t.Run(fmt.Sprintf("%s: Check expected resources are present", backend), CheckClusterExpectedResources)
+
+		t.Run(fmt.Sprintf("%s: Create a test table", backend), func(t *testing.T) {
+			CreateCassandraTable(t, medusaTestTable, medusaTestKeyspace)
```

Comment:
As with `TestReaperDeploymentScenario` I am not sure it makes sense to organize all of these as subtests. If `CreateCassandraTable` fails, then presumably all of the following tests that depend on that table will fail as well. You could conditionally execute the following tests only if `CreateCassandraTable` succeeds, but I introduces unnecessary complexity. If they are combined into a single test and `CreateCassandraTable` fails, then the test halts execution and no need for conditional logic.

Reply:
Yes, each subtest should be a "self-contained" test in a sense that it's a sequence of steps. So for example, these 8 steps could be one test. If you wish to have the logging, add t.Log() instead of t.Run(fmt.Sprintf( and so on.

Reply:
It could be more efficient to have a live chat about this because we don't seem to consider the same requirements in terms of parallelism and why it would be useful at this stage. If we don't have the same goal that conversation will quickly heat up as we talk past each other 😅

Reply:
@adejanovski what you explained to me about the CI set up mostly addressed my questions around the test organization/structure in the context of CI. To be clear though what I had envisioned for subtests was more along the lines of what @burmanm is describing.

 The explanation does not really address running tests locally. In a [previous comment](https://github.com/k8ssandra/k8ssandra/pull/544#discussion_r595794009) you gave the following example to demonstrate that the test execution will stop after the first failure:

```
% go test . -v -failfast -run=TestMedusaDeploymentScenario/"Minio:"
```

This is fine if I just want to run one test and its subtests. I don't think this works when running multiple tests. Test execution will stop on the first failure and not execute subsequent tests. It is nice to have the `-failfast` option, but when I run multiple tests, I want all of them to run and see the failures at the end.

I feel like the current approach with the fail fast, fined-grained subtests puts us back into a situation like we are in with Ginkgo, and that is a reason I wanted to move from Ginkgo to `testing`. 

On a semi-related note, gomega, at least as used, does not provide the behavior I want. Let me illustrate with examples. Here is a test using gomega:

```go
func TestFoo(t *testing.T) {
	g(t).Expect(1).To(gomega.Equal(2))
	g(t).Expect(1).To(gomega.Equal(3))
}
```
and here is the output when I run it:

```
$ go test -v -run=TestFoo ./tests/integration/...
=== RUN   TestFoo
    integration_suite_test.go:86:
        Expected
            <int>: 1
        to equal
            <int>: 2
--- FAIL: TestFoo (0.00s)
```
The second statement does not get executed. Now look at this next example:

```go
func TestBar(t *testing.T) {
	t.Error("1 != 2")
	t.Error("1 != 3")
}
```
and here is the output when I run it:

```
$ go test -v -run=TestBar ./tests/integration/...
=== RUN   TestBar
    integration_suite_test.go:91: 1 != 2
    integration_suite_test.go:92: 1 != 3
--- FAIL: TestBar (0.00s)
```
The errors are logged and execution continues until the end of the test. This is critical functionality IMO for integration/e2e tests.

Reply:
A single test is something that verifies functionality. These steps are often not verifying a functionality, but more like a single assertions only. "CreateCassandraTable" is not a test for k8ssandra, we don't ship "CassandraTableCreator". "Verify Medusa works" is a test, it's one of the functions we ship. The distinction is a bit more complex between "create a backup and verify it works" and "create a backup, verify it works and then restore it". Those I could see as two different tests.

Reply:
@burmanm, I find that definition a little dogmatic 🙂 . Anything that triggers an action and then performs checks through assertions could qualify as a test. In your example, creating a table allows us to check that Cassandra is reachable and behaves as expected. I guess the granularity is then a matter of personal taste and we need to commonly agree on the level we want.



Also, I'd like us to be on the same page: Wasn't the ability to run a single step one of your requirements? If we remove the subtests altogether and replace them with loggings, we'll only be able to run the whole scenario each time.

@jsanda, I don't think subtests are the main issue in the behavior you describe: it's the distinction between `Fail` and `Fatal` which can't be made when using Gomega. [It defaults](https://github.com/onsi/gomega/pull/215/files) to `Fatal` while the `testing` package provides both behaviors (`Fail` marking the test as failed but letting the execution continue). Your example with `t.Error()` isn't comparable because it calls `Log.Print()` and then `t.Fail()` underneath.
Now what we could do is use Gomega only for assertions that we consider fatal (or use Gomega only for its async assertions through `Eventually`).
Also both your snippets aren't equivalent as `t.Error()` is not performing an assertion, it's marking the running test case as failed without stopping its execution.

Then we can definitely discuss the organization of subtests and nest them accordingly to what is ok to run after a failed one. Let's take the following structure:
```
func TestReaperDeploymentScenario(t *testing.T) {
	ctx := initContext()
	t.Run("Initialize cluster", func(t *testing.T) {
		initializeCluster(t, ctx)
		t.Run("Deploy cluster", func(t *testing.T) {
			...
			...
			t.Run("Verify Reaper works as expected", func(t *testing.T) {
				t.Fatal("Fake failure")
				...
			})
			t.Run("Verify Prometheus is accessible", func(t *testing.T) {
				log.Println(Info("Prometheus is accessible, yay!"))
			})
			cleanupCluster(t, ctx)
		})
	})
}
```

Assuming we don't use `-failfast`, the `Verify Reaper works as expected` will fail but `Verify Prometheus is accessible` will be able to run:

```
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Reaper_works_as_expected
    integration_suite_test.go:110: Fake failure
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Prometheus_is_accessible
2021/03/18 10:05:59 Prometheus is accessible, yay!
2021/03/18 10:05:59 === Cleaning up cluster
TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster 2021-03-18T10:06:12+01:00 client.go:33: Configuring kubectl using config file /Users/adejanovski/.kube/config with context
TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster 2021-03-18T10:06:12+01:00 client.go:33: Configuring kubectl using config file /Users/adejanovski/.kube/config with context
TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster 2021-03-18T10:06:12+01:00 client.go:33: Configuring kubectl using config file /Users/adejanovski/.kube/config with context
--- FAIL: TestReaperDeploymentScenario (467.89s)
    --- FAIL: TestReaperDeploymentScenario/Initialize_cluster (467.89s)
        --- FAIL: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster (462.21s)
            --- FAIL: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Reaper_works_as_expected (0.00s)
            --- PASS: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Prometheus_is_accessible (0.00s)
```

But it will allow cleanupCluster() to run, which is not really something we want as it will prevent us from collecting artifacts or investigate directly on the k8s cluster.


That structure though will prevent any nested subtest from running if the parent test fails, which allows us to avoid running tests that would fail for sure.
If I move the failure at the beginning of the `Deploy cluster` subtest, I can see that no nested test will run:

```
func TestReaperDeploymentScenario(t *testing.T) {
	ctx := initContext()
	t.Run("Initialize cluster", func(t *testing.T) {
		initializeCluster(t, ctx)
		t.Run("Deploy cluster", func(t *testing.T) {
			t.Fatal("Fake failure")
			...
			t.Run("Verify Reaper works as expected", func(t *testing.T) {
				...
			})
			t.Run("Verify Prometheus is accessible", func(t *testing.T) {
				log.Println(Info("Prometheus is accessible, yay!"))
			})
			cleanupCluster(t, ctx)
		})
	})
}
```
It gives the following execution:
```
=== RUN   TestReaperDeploymentScenario/Initialize_cluster
...
...
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster
    integration_suite_test.go:105: Fake failure
--- FAIL: TestReaperDeploymentScenario (5.76s)
    --- FAIL: TestReaperDeploymentScenario/Initialize_cluster (5.76s)
        --- FAIL: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster (0.00s)
```

Now, `t.Run()` returns a boolean which tells us if the subtests succeeded. We could do the following to only cleanup the cluster if both the lowest level tests succeeded:
```
func TestReaperDeploymentScenario(t *testing.T) {
	ctx := initContext()
	t.Run("Initialize cluster", func(t *testing.T) {
		initializeCluster(t, ctx)
		t.Run("Deploy cluster", func(t *testing.T) {
			...
			reaperSucceeded := t.Run("Verify Reaper works as expected", func(t *testing.T) {
				t.Fatal("Fake failure")
				...
			})
			promSucceeded := t.Run("Verify Prometheus is accessible", func(t *testing.T) {
				log.Println(Info("Prometheus is accessible, yay!"))
			})
			if reaperSucceeded && promSucceeded {
				t.Run("Cleanup cluster", func(t *testing.T) {
					cleanupCluster(t, ctx)
				})
			}
		})
	})
}
```

Which gives the following output, effectively skipping the cleanup cluster test:
```
=== RUN   TestReaperDeploymentScenario/Initialize_cluster
...
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster
...
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Reaper_works_as_expected
    integration_suite_test.go:110: Fake failure
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Prometheus_is_accessible
2021/03/18 10:46:32 Prometheus is accessible, yay!
--- FAIL: TestReaperDeploymentScenario (475.83s)
    --- FAIL: TestReaperDeploymentScenario/Initialize_cluster (475.83s)
        --- FAIL: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster (470.12s)
            --- FAIL: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Reaper_works_as_expected (0.00s)
            --- PASS: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Prometheus_is_accessible (0.00s)
```

The above code would provide, I think, the behavior that you want, which could be further extended by not using Gomega for non fatal checks (although I don't really get the point of such assertions):

- Remove the use of `-failfast`
- Allow to run subtests from the same parent subtest to run despite each other success
- Prevent nested subtest to run if the parent test failed (giving us an explicit failfast behavior) and allow conditional behavior on running tests at the same level.

What we lose though:

- A bit of readability and simplicity of the code (the flat structure is easier to read)
- The ability to run individual subtests if they are nested ones. All the parent tests will have to run in order for the target subtest to run. That pretty much brings us back to Ginkgo's "run all or nothing":
```
% go test  -test.timeout=30m ./tests/integration -v -run=TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/"Verify Prometheus"
=== RUN   TestReaperDeploymentScenario
=== RUN   TestReaperDeploymentScenario/Initialize_cluster
...
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster
...
=== RUN   TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Prometheus_is_accessible
2021/03/18 11:06:10 Prometheus is accessible, yay!
--- PASS: TestReaperDeploymentScenario (456.56s)
    --- PASS: TestReaperDeploymentScenario/Initialize_cluster (456.56s)
        --- PASS: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster (450.84s)
            --- PASS: TestReaperDeploymentScenario/Initialize_cluster/Deploy_cluster/Verify_Prometheus_is_accessible (0.00s)
```

Reply:
Another option could be to avoid nesting subtests and rely on testing the result of mandatory ones:
```
func TestReaperDeploymentScenario(t *testing.T) {
	ctx := initContext()
	t.Run("Initialize cluster", func(t *testing.T) {
		initializeCluster(t, ctx)
	})

	deploySucceeded := t.Run("Deploy cluster", func(t *testing.T) {
		deployClusterForReaper(t, ctx)
		checkResourcePresenceForReaper(t, ctx)
	})

	if deploySucceeded {
		t.Run("Wait for Reaper pod to be ready", func(t *testing.T) {
			waitForReaperPod(t, ctx)
		})
		t.Run("Check Reaper has registered the required elements", func(t *testing.T) {
			checkReaperRegistered(t, ctx)
		})
		repairStarted := t.Run("Start repair on the reaper_db keyspace", func(t *testing.T) {
			triggerRepair(t, ctx)
		})

		if repairStarted {
			repairSuccess := t.Run("Wait for at least one segment to be processed and cancel the repair", func(t *testing.T) {
				waitForSegmentDoneAndCancel(t, ctx)
			})

			if repairSuccess {
				t.Run("Tear down cluster", func(t *testing.T) {
					cleanupCluster(t, ctx)
				})
			}
		}
	}
}
```
This allows us to run a single subtest even if it's conditioned to a subtests that won't be executed:
```
% go test  -test.timeout=30m ./tests/integration -v -run=TestReaperDeploymentScenario/"Start repair on the reaper_db keyspace"
=== RUN   TestReaperDeploymentScenario
=== RUN   TestReaperDeploymentScenario/Start_repair_on_the_reaper_db_keyspace
...
2021/03/18 11:12:13 Reaper response:
2021/03/18 11:12:13 Reaper status code: 200
--- PASS: TestReaperDeploymentScenario (0.69s)
    --- PASS: TestReaperDeploymentScenario/Start_repair_on_the_reaper_db_keyspace (0.69s)
PASS
ok  	github.com/k8ssandra/k8ssandra/tests/integration	1.138s
```
We can check how it behaves when triggering a Fatal failure on a subtest that is not conditioning subsequent subtests and then one one that is:
```
func TestReaperDeploymentScenario(t *testing.T) {
	ctx := initContext()
	t.Run("Initialize cluster", func(t *testing.T) {
		t.Fatal("Not really fatal failure")
		initializeCluster(t, ctx)
	})

	deploySucceeded := t.Run("Deploy cluster", func(t *testing.T) {
		t.Fatal("Really fatal failure")
		deployClusterForReaper(t, ctx)
		checkResourcePresenceForReaper(t, ctx)
	})

	if deploySucceeded {
		t.Run("Wait for Reaper pod to be ready", func(t *testing.T) {
			waitForReaperPod(t, ctx)
		})
...

```
Gives:
```
% grc go test  -test.timeout=30m ./tests/integration -v -run=TestReaperDeploymentScenario
=== RUN   TestReaperDeploymentScenario
=== RUN   TestReaperDeploymentScenario/Initialize_cluster
    integration_suite_test.go:72: Non fatal failure
=== RUN   TestReaperDeploymentScenario/Deploy_cluster
    integration_suite_test.go:77: Really fatal failure
--- FAIL: TestReaperDeploymentScenario (0.00s)
    --- FAIL: TestReaperDeploymentScenario/Initialize_cluster (0.00s)
    --- FAIL: TestReaperDeploymentScenario/Deploy_cluster (0.00s)
FAIL
FAIL	github.com/k8ssandra/k8ssandra/tests/integration	0.333s
FAIL
```

Would that better match your requirements?

Reply:
Testing that Cassandra can write to a table is more like a setup phase, or at most a unit test. It's not an integration test since there's nothing that was integrated - not Medusa and Cassandra. Nor a e2e test since we didn't actually accomplish anything.

I don't find anything useful in writing to a Cassandra table, why would I want to test that? And if I do, I'd like to see it verify the structure of the Cassandra files in the memtable, disk etc. What functionality would it test in k8ssandra? And if there's nothing useful in an assertion, then it's not a test - it's just a step inside a test. Same with some Medusa step, I don't want to test that its s3 integration can list a directory - that's a unit test. I want to test that it can create a backup from my Cassandra installation and send it to a S3, that's 3 components that are integrated and work together - that's what I want to test. In Medusa's unit tests you can verify the internal behavior only.

In the Reaper example:

```go
	if deploySucceeded {
		t.Run("Wait for Reaper pod to be ready", func(t *testing.T) {
			waitForReaperPod(t, ctx)
		})
		t.Run("Check Reaper has registered the required elements", func(t *testing.T) {
			checkReaperRegistered(t, ctx)
		})
		repairStarted := t.Run("Start repair on the reaper_db keyspace", func(t *testing.T) {
			triggerRepair(t, ctx)
		})
```

This should rather be:

```go
t.Log("Wait for Reaper pod to be ready")
waitForReaperPod(t, ctx)
t.Log("Check Reaper has registered the required elements")
checkReaperRegistered(t, ctx)
t.Logf(""Start repair on the %s keyspace", reaperKeyspace)
triggerRepair(t, ctx)
```

Although I would probably bundle the first two functions together, unless you want to test something that hasn't deployed reaper. And if that's the case, then the ``checkReaperRegistered`` should take a parameter what reaper and what cluster are registered. At the moment it just magically checks something that's hardcoded (at least for the reader) - nothing reusable and should not in that sense be a separate function. Making it a function for a struct makes it again not-reusable, so I don't think that's the correct choice - it should just take the input parameters. A massive parameter struct doesn't look nice either, do each of the functions really need all that information? I'd rather see it from my IDE which input parameters it needs and use the return values like they're meant to be used - it's more like Go in that case (it's in the effective go to avoid mutable input parameter style as the output). 

Think of it like:

```go
t.Log("Deploying a reaper instance")
reaper := deployReaperInstance(t, namespace)
t.Log("Wait for Reaper pod to be ready")
waitForReaperPod(t, reaper)
t.Log("Check Reaper has registered the required elements")
checkReaperRegistered(t, reaper, cassdc)
t.Logf(""Start repair on the %s keyspace", reaperKeyspace)
triggerRepair(t, reaper, reaperKeyspace)
```

Or even removing the whole t from those and doing the assertion here (I'll write with gomega, but that's not relevant here):

```go
t.Log("Deploying a reaper instance")
reaper, err := deployReaperInstance(namespace)
Expect(err).To(BeNil())

t.Log("Wait for Reaper pod to be ready")
Expect(waitForReaperPod(reaper)).To(Succeed())

t.Log("Check Reaper has registered the required elements")
err = checkReaperRegistered(reaper, cassdc)
Expect(err).To(BeNil()) // check for k8sClient failures

// Or, if the checks are simple - keep them inline since too small functions are not worthy of being functions
Expect(reaper.Clusters).To(HaveKey(cassdc))

t.Logf("Start repair on the %s keyspace", reaperKeyspace)
Expect(triggerRepair(reaper, reaperKeyspace)).To(Succeed())
```

Now, with a single glance I can see the steps the test took and I can see all the verifications it did. I simplified my methods and I made sure my methods do not have to be test-only, lets say some of them called a client library such as reaper-client-go. I've reused existing code or I've allowed some other functions in the software to reuse what I've written. And if my code and test use the same methods to deploy the reaper - even better, then I'm really testing the actual implementation. Rule of thumb, if the test can fit a single screen, it's not too complex and should not be split.

Reply:
> Testing that Cassandra can write to a table is more like a setup phase

I still disagree and think this is semantics. Even then do you really see this as a blocker for this PR?
Maybe the code currently splits up in too many steps (depending on your sensitivity) that could be bundled together. But it brings some clarity which will be helpful to understand what these tests do instead of having to follow the trail of all that's being done by following the track of each sub function call.
We can always be more concise and wrap everything together, but that complicates understanding what the test does.

Same goes for the output: subtests allow us to have a very granular view of which tests passed and which failed in the summary:

```
--- FAIL: TestReaperDeploymentScenario (0.00s)
    --- FAIL: TestReaperDeploymentScenario/Initialize_cluster (0.00s)
    --- PASS: TestReaperDeploymentScenario/Test_that_will_pass_for_sure (0.00s)
    --- FAIL: TestReaperDeploymentScenario/Deploy_cluster (0.00s)
```
We won't get that if we don't use subtests.

Let's put aside the input parameters and hardcoded stuff you're referring to in order to focus on flat vs subtests. The context variable introduction is a wip to remove global variables as per one of your comments and based on a suggestion from John. This integration test suite will hopefully go through refactorings in the future and evolve in a direction or another depending on the needs we have.

Mixing concerns makes it harder to answer the initial questions, so I'd like to refocus on the initial request from @jsanda's: have a way for several checks to fail without preventing other specific checks from running, while still allowing to stop the test execution as other subsequent checks would consistently fail. So `-failfast` was too extreme and he wanted more granular control on what can run.
How would you solve this in your proposal?

Also you repeatedly dodge the question about being able to run only specific steps in a scenario: Is that a requirement or do you want to re-run the whole scenario each time one step fails?

Reply:
I have no interest in running each assertion, I can't do that in real life. Lets say I modified the reaper-operator and before pushing I want to test it.

What I need is to run the "setup phase" first to setup my cluster. What I then want is to run the reaper process and let it clean the mess, fix my errors, rerun the reaper process and fix the mess. I don't want to run smaller parts - because that testing should be part of reaper-operator. I also can't run smaller subtests, because I will probably need to cleanup the mess that might have been caused by the buggy operator for example. So I need the init and clean for that reaper data, but I don't need to clean and redo the CassandraDatacenter part - I might need to remove an annotation, but that was hopefully something the reaper part added and removed.

We need to keep in mind what tests are part of k8ssandra and which are part of the individual operators. Just like we're not copying Medusa's internal tests here, we only want to see it from the sky-high perspective, higher than what medusa-operator does. We're not even testing medusa-operator's single details. And whatever this higher view is - is a single test which might need running. If it's smaller test, then it should be done in the component itself and not here.

Reply:
>I don't think subtests are the main issue in the behavior you describe: it's the distinction between Fail and Fatal which can't be made when using Gomega. It defaults to Fatal while the testing package provides both behaviors (Fail marking the test as failed but letting the execution continue). 

Subtests as designed particularly with `-failfast` are my main concern. I am also pointing out the distinction between `Fail`/`Error` and `Fatal` which is an important one.

>Also both your snippets aren't equivalent as t.Error() is not performing an assertion, it's marking the running test case as failed without stopping its execution.

The snippets were not intended to be the same. They were intended to contrast the differences.

Reply:
>Assuming we don't use -failfast, the Verify Reaper works as expected will fail but Verify Prometheus is accessible will be able to run:
...
>But it will allow cleanupCluster() to run, which is not really something we want as it will prevent us from collecting artifacts or investigate directly on the k8s cluster.

We can easily track whether there are any subtest failures and check that in `cleanupCluster`.

Reply:
>The above code would provide, I think, the behavior that you want, which could be further extended by not using Gomega for non fatal checks (although I don't really get the point of such assertions):
>* Remove the use of -failfast
>* Allow to run subtests from the same parent subtest to run despite each other success
>* Prevent nested subtest to run if the parent test failed (giving us an explicit failfast behavior) and allow conditional behavior on running tests at the same level.

I think we are making this more complicated than it needs to be right now. Do we really need to decide how/when/if all subtests will run right now? 

If I write a test with subtests A, B, and C and I want to skip B if A fails but still run C, then I can add the logic to do so. Over time if we see common patterns then we refactor accordingly.

Reply:
Here is what I have had in mind for subtest organization:

```go
func TestInstallFullStack(t* testing.T) {
    t.Run("verify reaper stuff", testReaperInstall)
    t.Run("verify stargate install", testStargateInstall)
    t.Run("verify medusa install", testMedusaInstall)
    t.Run("verify prometheus install", testPrometheusInstall)
    t.Run("verify grafana install", testGrafanaInstall) 
}
```
I don't think we need subtests to be more fine-grained than this. Of course this is pretty subjective.